File: scripts/offline_assets.py
#!/usr/bin/env python3
"""
TrackerClaw — Offline / Manual Asset Manager
Manages non-crypto assets (cash, stocks, gold, etc.) in data/offline.json.

Usage:
    python offline_assets.py list
    python offline_assets.py add cash 5000 USD
    python offline_assets.py add stock AAPL 12 shares 220 avg
    python offline_assets.py update <id> price 245
    python offline_assets.py remove <id>
    python offline_assets.py --json
"""

import json
import os
import sys
import argparse
import time
from datetime import datetime, timezone

import requests

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_DIR = os.path.dirname(SCRIPT_DIR)
OFFLINE_FILE = os.path.join(PROJECT_DIR, "data", "offline.json")

# Yahoo Finance v8 API (public, no key needed)
YAHOO_QUOTE_URL = "https://query1.finance.yahoo.com/v8/finance/chart/{symbol}?range=1d&interval=1d"


def load_offline() -> list:
    """Load offline assets from JSON file."""
    os.makedirs(os.path.dirname(OFFLINE_FILE), exist_ok=True)
    if os.path.exists(OFFLINE_FILE):
        with open(OFFLINE_FILE, "r") as f:
            data = json.load(f)
            return data if isinstance(data, list) else []
    return []


def save_offline(assets: list):
    """Save offline assets to JSON file."""
    os.makedirs(os.path.dirname(OFFLINE_FILE), exist_ok=True)
    with open(OFFLINE_FILE, "w") as f:
        json.dump(assets, f, indent=2)


def next_id(assets: list) -> int:
    """Get next available ID."""
    if not assets:
        return 1
    return max(a.get("id", 0) for a in assets) + 1


def fetch_stock_price(symbol: str) -> float:
    """Fetch current stock price from Yahoo Finance."""
    try:
        import yfinance as yf
        ticker = yf.Ticker(symbol)
        info = ticker.fast_info
        return float(info.get("lastPrice", 0) or info.get("last_price", 0))
    except Exception:
        pass

    # Fallback: direct API call
    try:
        url = YAHOO_QUOTE_URL.format(symbol=symbol.upper())
        headers = {"User-Agent": "Mozilla/5.0"}
        resp = requests.get(url, headers=headers, timeout=10)
        if resp.status_code == 200:
            data = resp.json()
            result = data.get("chart", {}).get("result", [])
            if result:
                meta = result[0].get("meta", {})
                return float(meta.get("regularMarketPrice", 0))
    except Exception:
        pass

    return 0.0


def cmd_list(as_json: bool = False):
    """List all offline assets."""
    assets = load_offline()

    if not assets:
        print("No offline assets tracked. Use 'add' to add some.")
        return

    # Update stock prices
    for a in assets:
        if a.get("type") == "stock" and a.get("symbol"):
            print(f"⏳ Fetching price for {a['symbol']}...")
            current = fetch_stock_price(a["symbol"])
            if current > 0:
                a["current_price"] = current
                a["current_value"] = current * a.get("shares", 0)

    save_offline(assets)

    if as_json:
        print(json.dumps(assets, indent=2))
        return

    total = 0.0
    print("**TrackerClaw — Offline Assets**\n")
    for a in assets:
        asset_id = a.get("id", "?")
        atype = a.get("type", "unknown")

        if atype == "cash":
            amount = a.get("amount", 0)
            currency = a.get("currency", "USD")
            total += amount
            print(f"• [#{asset_id}] Cash — {currency} {amount:,.2f}")

        elif atype == "stock":
            symbol = a.get("symbol", "???")
            shares = a.get("shares", 0)
            avg_price = a.get("avg_price", 0)
            current = a.get("current_price", avg_price)
            value = current * shares
            total += value
            pnl = (current - avg_price) * shares if avg_price > 0 else 0
            pnl_pct = ((current / avg_price) - 1) * 100 if avg_price > 0 else 0
            pnl_sign = "+" if pnl >= 0 else ""
            print(f"• [#{asset_id}] {symbol} — {shares} shares @ ${current:,.2f} "
                  f"= ${value:,.2f} ({pnl_sign}${pnl:,.2f} / {pnl_sign}{pnl_pct:.1f}%)")

        elif atype == "commodity":
            name = a.get("name", "Unknown")
            amount = a.get("amount", 0)
            unit = a.get("unit", "units")
            price = a.get("price_per_unit", 0)
            value = amount * price
            total += value
            print(f"• [#{asset_id}] {name} — {amount} {unit} @ ${price:,.2f} = ${value:,.2f}")

        else:
            name = a.get("name", atype)
            value = a.get("value", 0)
            total += value
            print(f"• [#{asset_id}] {name} — ${value:,.2f}")

    print(f"\n**Offline Total: ${total:,.2f}**")


def cmd_add(args: list):
    """Add an offline asset."""
    if len(args) < 3:
        print("Usage:")
        print("  add cash <amount> <currency>")
        print("  add stock <symbol> <shares> shares <avg_price> avg")
        print("  add commodity <name> <amount> <unit> <price_per_unit>")
        print("  add other <name> <value>")
        return

    assets = load_offline()
    asset_type = args[0].lower()
    aid = next_id(assets)

    if asset_type == "cash":
        amount = float(args[1])
        currency = args[2].upper() if len(args) > 2 else "USD"
        asset = {
            "id": aid,
            "type": "cash",
            "amount": amount,
            "currency": currency,
            "added": datetime.now(timezone.utc).isoformat()
        }
        assets.append(asset)
        save_offline(assets)
        print(f"✅ Added cash: {currency} {amount:,.2f} (#{aid})")

    elif asset_type == "stock":
        symbol = args[1].upper()
        shares = float(args[2])
        avg_price = 0.0
        # Parse optional "shares" and "avg" keywords
        remaining = args[3:]
        for i, a in enumerate(remaining):
            if a.lower() == "avg" and i > 0:
                try:
                    avg_price = float(remaining[i - 1])
                except ValueError:
                    pass
            elif a.lower() not in ("shares", "avg"):
                try:
                    avg_price = float(a)
                except ValueError:
                    pass

        # Fetch current price
        current_price = fetch_stock_price(symbol)
        if current_price == 0 and avg_price > 0:
            current_price = avg_price

        asset = {
            "id": aid,
            "type": "stock",
            "symbol": symbol,
            "shares": shares,
            "avg_price": avg_price,
            "current_price": current_price,
            "current_value": current_price * shares,
            "added": datetime.now(timezone.utc).isoformat()
        }
        assets.append(asset)
        save_offline(assets)
        print(f"✅ Added stock: {symbol} × {shares} @ avg ${avg_price:,.2f} "
              f"(current: ${current_price:,.2f}) (#{aid})")

    elif asset_type == "commodity":
        name = args[1]
        amount = float(args[2])
        unit = args[3] if len(args) > 3 else "units"
        price = float(args[4]) if len(args) > 4 else 0.0
        asset = {
            "id": aid,
            "type": "commodity",
            "name": name,
            "amount": amount,
            "unit": unit,
            "price_per_unit": price,
            "added": datetime.now(timezone.utc).isoformat()
        }
        assets.append(asset)
        save_offline(assets)
        print(f"✅ Added commodity: {name} × {amount} {unit} @ ${price:,.2f} (#{aid})")

    elif asset_type == "other":
        name = args[1]
        value = float(args[2]) if len(args) > 2 else 0.0
        asset = {
            "id": aid,
            "type": "other",
            "name": name,
            "value": value,
            "added": datetime.now(timezone.utc).isoformat()
        }
        assets.append(asset)
        save_offline(assets)
        print(f"✅ Added: {name} — ${value:,.2f} (#{aid})")

    else:
        print(f"❌ Unknown asset type: {asset_type}")
        print("Supported types: cash, stock, commodity, other")


def cmd_update(asset_id: str, field: str, value: str):
    """Update an offline asset field."""
    assets = load_offline()
    try:
        aid = int(asset_id)
    except ValueError:
        print(f"❌ Invalid asset ID: {asset_id}")
        return

    found = False
    for a in assets:
        if a.get("id") == aid:
            found = True
            if field.lower() == "price":
                new_price = float(value)
                if a.get("type") == "stock":
                    a["current_price"] = new_price
                    a["current_value"] = new_price * a.get("shares", 0)
                    print(f"✅ Updated #{aid} price to ${new_price:,.2f}")
                elif a.get("type") == "commodity":
                    a["price_per_unit"] = new_price
                    print(f"✅ Updated #{aid} price to ${new_price:,.2f}")
                else:
                    a["value"] = new_price
                    print(f"✅ Updated #{aid} value to ${new_price:,.2f}")
            elif field.lower() == "amount":
                a["amount"] = float(value)
                print(f"✅ Updated #{aid} amount to {value}")
            elif field.lower() == "shares":
                a["shares"] = float(value)
                print(f"✅ Updated #{aid} shares to {value}")
            else:
                a[field] = value
                print(f"✅ Updated #{aid} {field} to {value}")
            a["updated"] = datetime.now(timezone.utc).isoformat()
            break

    if not found:
        print(f"❌ No asset found with ID #{asset_id}")
        return

    save_offline(assets)


def cmd_remove(asset_id: str):
    """Remove an offline asset."""
    assets = load_offline()
    try:
        aid = int(asset_id)
    except ValueError:
        print(f"❌ Invalid asset ID: {asset_id}")
        return

    original_len = len(assets)
    assets = [a for a in assets if a.get("id") != aid]

    if len(assets) == original_len:
        print(f"❌ No asset found with ID #{asset_id}")
        return

    save_offline(assets)
    print(f"✅ Removed asset #{asset_id}")


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    command = sys.argv[1].lower()

    if command == "list":
        as_json = "--json" in sys.argv or "-j" in sys.argv
        cmd_list(as_json=as_json)
    elif command == "add":
        cmd_add(sys.argv[2:])
    elif command == "update":
        if len(sys.argv) < 5:
            print("Usage: python offline_assets.py update <id> <field> <value>")
            sys.exit(1)
        cmd_update(sys.argv[2], sys.argv[3], sys.argv[4])
    elif command == "remove":
        if len(sys.argv) < 3:
            print("Usage: python offline_assets.py remove <id>")
            sys.exit(1)
        cmd_remove(sys.argv[2])
    elif command == "--json" or command == "-j":
        cmd_list(as_json=True)
    else:
        print(f"Unknown command: {command}")
        print(__doc__)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scriptsTS/offline_assets.ts
#!/usr/bin/env node

import fs from "node:fs/promises";

import { DATA_DIR, OFFLINE_FILE } from "./lib/paths.js";
import { isoNow, writeJsonFile } from "./lib/utils.js";

const YAHOO_QUOTE_URL = "https://query1.finance.yahoo.com/v8/finance/chart";

interface OfflineAsset {
  id: number;
  type: string;
  amount?: number;
  currency?: string;
  symbol?: string;
  shares?: number;
  avg_price?: number;
  current_price?: number;
  current_value?: number;
  name?: string;
  unit?: string;
  price_per_unit?: number;
  value?: number;
  added?: string;
  updated?: string;
}

export async function loadOffline(): Promise<OfflineAsset[]> {
  await fs.mkdir(DATA_DIR, { recursive: true });
  try {
    const raw = await fs.readFile(OFFLINE_FILE, "utf8");
    const payload = JSON.parse(raw);
    return Array.isArray(payload) ? payload : [];
  } catch {
    return [];
  }
}

export async function saveOffline(assets: OfflineAsset[]): Promise<void> {
  await fs.mkdir(DATA_DIR, { recursive: true });
  await writeJsonFile(OFFLINE_FILE, assets);
}

function nextId(assets: OfflineAsset[]): number {
  return assets.length ? Math.max(...assets.map((asset) => asset.id || 0)) + 1 : 1;
}

export async function fetchStockPrice(symbol: string): Promise<number> {
  try {
    const url = new URL(`${YAHOO_QUOTE_URL}/${symbol.toUpperCase()}`);
    url.searchParams.set("range", "1d");
    url.searchParams.set("interval", "1d");
    const response = await fetch(url, {
      headers: { "user-agent": "Mozilla/5.0" },
    });
    if (!response.ok) {
      return 0;
    }
    const data = (await response.json()) as any;
    return Number(data.chart?.result?.[0]?.meta?.regularMarketPrice || 0);
  } catch {
    return 0;
  }
}

export async function cmdList(asJson = false): Promise<void> {
  const assets = await loadOffline();
  if (!assets.length) {
    console.log("No offline assets tracked. Use 'add' to add some.");
    return;
  }

  for (const asset of assets) {
    if (asset.type === "stock" && asset.symbol) {
      const current = await fetchStockPrice(asset.symbol);
      if (current > 0) {
        asset.current_price = current;
        asset.current_value = current * Number(asset.shares || 0);
      }
    }
  }
  await saveOffline(assets);

  if (asJson) {
    console.log(JSON.stringify(assets, null, 2));
    return;
  }

  let total = 0;
  console.log("**TrackerClaw — Offline Assets**\n");
  for (const asset of assets) {
    if (asset.type === "cash") {
      total += Number(asset.amount || 0);
      console.log(`• [#${asset.id}] Cash — ${asset.currency || "USD"} ${Number(asset.amount || 0).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`);
      continue;
    }
    if (asset.type === "stock") {
      const shares = Number(asset.shares || 0);
      const avgPrice = Number(asset.avg_price || 0);
      const current = Number(asset.current_price || avgPrice);
      const value = current * shares;
      total += value;
      const pnl = avgPrice > 0 ? (current - avgPrice) * shares : 0;
      const pnlPct = avgPrice > 0 ? ((current / avgPrice) - 1) * 100 : 0;
      const sign = pnl >= 0 ? "+" : "";
      console.log(
        `• [#${asset.id}] ${asset.symbol} — ${shares} shares @ $${current.toFixed(2)} = $${value.toFixed(2)} (${sign}$${pnl.toFixed(2)} / ${sign}${pnlPct.toFixed(1)}%)`,
      );
      continue;
    }
    if (asset.type === "commodity") {
      const amount = Number(asset.amount || 0);
      const price = Number(asset.price_per_unit || 0);
      const value = amount * price;
      total += value;
      console.log(`• [#${asset.id}] ${asset.name} — ${amount} ${asset.unit || "units"} @ $${price.toFixed(2)} = $${value.toFixed(2)}`);
      continue;
    }
    const value = Number(asset.value || 0);
    total += value;
    console.log(`• [#${asset.id}] ${asset.name || asset.type} — $${value.toFixed(2)}`);
  }

  console.log(`\n**Offline Total: $${total.toFixed(2)}**`);
}

export async function cmdAdd(args: string[]): Promise<void> {
  const assets = await loadOffline();
  const assetType = (args[0] || "").toLowerCase();
  const id = nextId(assets);

  if (assetType === "cash") {
    const amount = Number(args[1] || 0);
    const currency = (args[2] || "USD").toUpperCase();
    assets.push({ id, type: "cash", amount, currency, added: isoNow() });
    await saveOffline(assets);
    console.log(`Added cash: ${currency} ${amount.toFixed(2)} (#${id})`);
    return;
  }

  if (assetType === "stock") {
    const symbol = (args[1] || "").toUpperCase();
    const shares = Number(args[2] || 0);
    const avgPrice = Number(args.find((value, index) => args[index + 1]?.toLowerCase() === "avg") || 0) || Number(args[4] || 0);
    const currentPrice = (await fetchStockPrice(symbol)) || avgPrice;
    assets.push({
      id,
      type: "stock",
      symbol,
      shares,
      avg_price: avgPrice,
      current_price: currentPrice,
      current_value: currentPrice * shares,
      added: isoNow(),
    });
    await saveOffline(assets);
    console.log(`Added stock: ${symbol} x ${shares} @ avg $${avgPrice.toFixed(2)} (current: $${currentPrice.toFixed(2)}) (#${id})`);
    return;
  }

  if (assetType === "commodity") {
    const name = args[1];
    const amount = Number(args[2] || 0);
    const unit = args[3] || "units";
    const price = Number(args[4] || 0);
    assets.push({
      id,
      type: "commodity",
      name,
      amount,
      unit,
      price_per_unit: price,
      added: isoNow(),
    });
    await saveOffline(assets);
    console.log(`Added commodity: ${name} x ${amount} ${unit} @ $${price.toFixed(2)} (#${id})`);
    return;
  }

  if (assetType === "other") {
    const name = args[1];
    const value = Number(args[2] || 0);
    assets.push({
      id,
      type: "other",
      name,
      value,
      added: isoNow(),
    });
    await saveOffline(assets);
    console.log(`Added: ${name} — $${value.toFixed(2)} (#${id})`);
    return;
  }

  throw new Error(`Unknown asset type: ${assetType}`);
}

export async function cmdUpdate(assetId: string, field: string, value: string): Promise<void> {
  const assets = await loadOffline();
  const id = Number(assetId);
  const asset = assets.find((entry) => entry.id === id);
  if (!asset) {
    throw new Error(`No asset found with ID #${assetId}`);
  }

  if (field.toLowerCase() === "price") {
    const price = Number(value);
    if (asset.type === "stock") {
      asset.current_price = price;
      asset.current_value = price * Number(asset.shares || 0);
    } else if (asset.type === "commodity") {
      asset.price_per_unit = price;
    } else {
      asset.value = price;
    }
  } else if (field.toLowerCase() === "amount") {
    asset.amount = Number(value);
  } else if (field.toLowerCase() === "shares") {
    asset.shares = Number(value);
  } else {
    (asset as unknown as Record<string, unknown>)[field] = value;
  }
  asset.updated = isoNow();
  await saveOffline(assets);
  console.log(`Updated #${assetId} ${field} to ${value}`);
}

export async function cmdRemove(assetId: string): Promise<void> {
  const assets = await loadOffline();
  const id = Number(assetId);
  const filtered = assets.filter((asset) => asset.id !== id);
  if (filtered.length === assets.length) {
    throw new Error(`No asset found with ID #${assetId}`);
  }
  await saveOffline(filtered);
  console.log(`Removed asset #${assetId}`);
}

async function main(): Promise<void> {
  const [command, ...args] = process.argv.slice(2);
  try {
    if (command === "list") {
      await cmdList(args.includes("--json") || args.includes("-j"));
      return;
    }
    if (command === "add") {
      await cmdAdd(args);
      return;
    }
    if (command === "update") {
      if (args.length < 3) {
        throw new Error("Usage: update <id> <field> <value>");
      }
      await cmdUpdate(args[0], args[1], args[2]);
      return;
    }
    if (command === "remove") {
      if (!args[0]) {
        throw new Error("Usage: remove <id>");
      }
      await cmdRemove(args[0]);
      return;
    }
    if (command === "--json" || command === "-j") {
      await cmdList(true);
      return;
    }
    throw new Error("Usage: tsx scriptsTS/offline_assets.ts list|add|update|remove");
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

if (process.argv[1] && process.argv[1].endsWith("offline_assets.ts")) {
  void main();
}
